server/app/middlewares/websocket_backend.rb

```diff
@@ -66,43 +87,84 @@ def call(env)
-      end
+    node = grid.host_nodes.find_by(node_id: node_id)
 
-      node_plugger = Agent::NodePlugger.new(grid, node)
-      client = {
-          ws: ws,
-          id: node_id.to_s,
-          node_id: node.id,
-          grid_id: grid.id,
-          created_at: Time.now
-      }
-      @clients << client
-
-      agent_version = req.env['HTTP_KONTENA_VERSION'].to_s
-      unless self.valid_agent_version?(agent_version)
-        logger.error "version mismatch: server (#{Server::VERSION}), node (#{agent_version})"
-        node_plugger.send_master_info
-        self.handle_invalid_agent_version(ws, node)
-        return
-      end
+    unless node
+      node = grid.host_nodes.create!(node_id: node_id, labels: labels)
+    end
 
-      EM.defer { node_plugger.plugin! }
-    else
-      logger.error 'invalid grid token, closing connection'
-      ws.close(4001)
+    # check version
+    agent_version = req.env['HTTP_KONTENA_VERSION'].to_s
+
+    unless self.valid_agent_version?(agent_version)
+      handle_invalid_agent_version(ws, node, agent_version)
+      return
     end
+
+    # check clock
+    connected_at = Time.parse(req.env['HTTP_KONTENA_CONNECTED_AT'])
+    connected_dt = Time.now - connected_at
+
+    if connected_dt > PING_TIMEOUT + CLOCK_SKEW || connected_dt < -CLOCK_SKEW
+      handle_invalid_agent_clock(ws, node, connected_at, connected_dt)
+      return
+    end
+
+    logger.info "node #{node} agent version #{agent_version} connected at #{connected_at}, #{'%.2fs' % connected_dt} ago"
+
+    client = {
+        ws: ws,
+        id: node_id.to_s,
+        node_id: node.id,
+        grid_id: grid.id,
+        created_at: Time.now,
+        connected_at: connected_at,
+    }
+
+    plugin_client(client, node)
   rescue => exc
-    logger.error "#{exc.class.name}: #{exc.message}"
+    logger.error exc
+  end
+
+  # @param [Faye::WebSocket] ws
+  # @param [HostNode] node
+  def handle_invalid_agent_version(ws, node, version)
+    logger.warn "node #{node} agent version #{version} is not compatible with server version #{Server::VERSION}"
+
+    ws.close(4010, "agent version #{version} is not compatible with server version #{Server::VERSION}")
+  end
+
+  def handle_invalid_agent_clock(ws, node, connected_at, connected_dt)
+    if connected_dt < 0
+      logger.warn "node #{node} connected too far in the future at #{connected_at}, #{'%.2fs' % -connected_dt} ahead"
+    else
+      logger.warn "node #{node} connected too far in the past at #{connected_at}, #{'%.2fs' % connected_dt} ago"
+    end
+    ws.close(4020, "agent clock offset #{'%.2fs' % connected_dt} exceeds threshold")
+  end
+
+  # Add to @clients, and mark client HostNode as connected
+  #
+  # @param [Hash] client
+  # @param [HostNode] node
+  def plugin_client(client, node)
+    @clients << client
+
+    node_plugger = Agent::NodePlugger.new(node, client[:connected_at])
+
+    EM.defer { node_plugger.plugin! }
```

Comment:
The `EM.defer { ... }` here is problematic, because if the `NodePlugger#update_node` detects a newer `connected_at`, then this should really fail the connection immediately, and not append it to `@clients` etc... but with the async call, that is difficult.